apis/v1beta1/config.go

```diff
@@ -445,67 +447,133 @@ const (
+	}
+
+	port := defaultServicePort
+	host := defaultServiceHost
+	for _, reader := range telemetry.Metrics.Readers {
+		if reader.Pull != nil {
+			if reader.Pull.Exporter.Prometheus != nil {
+				if reader.Pull.Exporter.Prometheus.Host != nil && *reader.Pull.Exporter.Prometheus.Host != "" {
+					host = *reader.Pull.Exporter.Prometheus.Host
+				}
+				if reader.Pull.Exporter.Prometheus.Port != nil && *reader.Pull.Exporter.Prometheus.Port != 0 {
+					port = intToInt32Safe(int(*reader.Pull.Exporter.Prometheus.Port))
+				}
+				break
+			}
+		}
+	}
+	return host, port, nil
 }
 
 // ApplyDefaults inserts configuration defaults if it has not been set.
 func (s *Service) ApplyDefaults(logger logr.Logger) error {
-	telemetryAddr, telemetryPort, err := s.MetricsEndpoint(logger)
-	if err != nil {
-		return err
+	tel := s.GetTelemetry()
+
+	if tel == nil {
+		tel = &Telemetry{}
 	}
+	s.configureMetricReaders(logger, tel)
 
-	tm := &AnyConfig{
-		Object: map[string]interface{}{
-			"metrics": map[string]interface{}{
-				"address": fmt.Sprintf("%s:%d", telemetryAddr, telemetryPort),
-			},
-		},
+	// This field is deprecated and should not be used anymore.
+	if tel.Metrics.Address != "" {
+		// Setting the "address" field to an empty string explicitly removes the value
+		// during Kubernetes serialization. Directly deleting the field from the map using
+		// delete(metrics, "address") does not work because Kubernetes treats missing fields
+		// differently from explicitly empty ones. By assigning "", we ensure the configuration
+		// is updated correctly when the resource is persisted.
+		tel.Metrics.Address = ""
 	}
 
-	if s.Telemetry == nil {
-		s.Telemetry = tm
-		return nil
+	yamlConfig, err := tel.toMap()
+	if err != nil {
+		return err
 	}
-	// NOTE: Merge without overwrite. If a telemetry endpoint is specified, the defaulting
-	// respects the configuration and returns an equal value.
-	if err := mergo.Merge(s.Telemetry, tm); err != nil {
-		return fmt.Errorf("telemetry config merge failed: %w", err)
+
+	s.Telemetry = &AnyConfig{
+		Object: yamlConfig,
 	}
+
 	return nil
 }
 
+func (s *Service) configureMetricReaders(logger logr.Logger, tel *Telemetry) {
+	port := defaultServicePort
+	host := defaultServiceHost
+	telemetryAddr, telemetryPort, err := s.MetricsEndpoint(logger)
+	if err != nil {
+		logger.Error(err, "couldn't determine metrics endpoint")
+	} else {
+		port = telemetryPort
+		host = telemetryAddr
+	}
+	// Port is an int32, but the MetricReader expects an int.
+	portInt := int(port)
+
+	if tel.Metrics.Readers == nil {
+		tel.Metrics.Readers = []otelConfig.MetricReader{
+			{
+				Pull: &otelConfig.PullMetricReader{
+					Exporter: otelConfig.PullMetricExporter{
+						Prometheus: &otelConfig.Prometheus{
+							Host: &host,
+							Port: &portInt,
+						},
+					},
+				},
+			},
+		}
+		return
+	}
+
+	for _, reader := range tel.Metrics.Readers {
+		if reader.Pull != nil {
+			if reader.Pull.Exporter.Prometheus != nil {
```

Comment:
is this a valid case? I guess i'm not sure if we need to do anything in the case someone has the pull exporter set to prometheus but no host/port. I think in this case, the host/port becomes the defaults [here](https://github.com/open-telemetry/opentelemetry-configuration/blob/2e76e05faf862650b23363d2a0f4d6b34df9e646/examples/kitchen-sink.yaml#L199-L202) and we wouldn't need to set this.

Reply:
In general, I don't think we should do anything to existing readers. This whole for loop is unnecessary. If the user set their own metric readers, we don't need to add any defaults ourselves.

Reply:
Ok, I will simplify this.